sql/src/main/java/io/crate/metadata/cluster/AlterTableClusterStateExecutor.java

```diff
@@ -0,0 +1,437 @@
+                indexMapperServices.put(index, mapperService);
+                // add mappings for all types, we need them for cross-type validation
+                mapperService.merge(indexMetaData, MapperService.MergeReason.MAPPING_RECOVERY, false);
+            }
+        }
+
+        PutMappingClusterStateUpdateRequest updateRequest = new PutMappingClusterStateUpdateRequest()
+            .ackTimeout(request.timeout()).masterNodeTimeout(request.masterNodeTimeout())
+            .indices(concreteIndices).type(Constants.DEFAULT_MAPPING_TYPE)
+            .source(request.source());
+
+        return metaDataMappingService.putMappingExecutor.applyRequest(currentState, updateRequest, indexMapperServices);
+    }
+
+    /**
+     * The logic is taken over from {@link MetaDataUpdateSettingsService#updateSettings(UpdateSettingsClusterStateUpdateRequest, ActionListener)}
+     */
+    private ClusterState updateSettings(final ClusterState currentState, final Settings settings, Index[] concreteIndices) {
+
+        final Settings normalizedSettings = Settings.builder().put(settings).normalizePrefix(IndexMetaData.INDEX_SETTING_PREFIX).build();
+        Settings.Builder settingsForClosedIndices = Settings.builder();
+        Settings.Builder settingsForOpenIndices = Settings.builder();
+        final Set<String> skippedSettings = new HashSet<>();
+
+        //TODO Do we need all the modifications of the settings ?
+        indexScopedSettings.validate(
+            normalizedSettings.filter(s -> Regex.isSimpleMatchPattern(s) == false), // don't validateSettings wildcards
+            false, // don't validateSettings dependencies here we check it below never allow to change the number of shards
+            true); // validateSettings internal or private index settings
+        for (String key : normalizedSettings.keySet()) {
+            Setting setting = indexScopedSettings.get(key);
+            boolean isWildcard = setting == null && Regex.isSimpleMatchPattern(key);
+            assert setting != null // we already validated the normalized settings
+                   || (isWildcard && normalizedSettings.hasValue(key) == false)
+                : "unknown setting: " + key + " isWildcard: " + isWildcard + " hasValue: " +
+                  normalizedSettings.hasValue(key);
+            settingsForClosedIndices.copy(key, normalizedSettings);
+            if (isWildcard || setting.isDynamic()) {
+                settingsForOpenIndices.copy(key, normalizedSettings);
+            } else {
+                skippedSettings.add(key);
+            }
+        }
+        final Settings closedSettings = settingsForClosedIndices.build();
+        final Settings openSettings = settingsForOpenIndices.build();
+        //final boolean preserveExisting = request.isPreserveExisting();
+        //TODO do we need this property ?
+        final boolean preserveExisting = false;
```

Comment:
I don't think so, new settings should always overwrite existing ones. But maybe I'm missing something here.